add configuration option for excluding certain requests from updating the user token

jeddai opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

* [x] The commit message follows our [guidelines](https://github.com/mauriciovigolo/keycloak-angular/blob/master/CONTRIBUTING.md)
* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

All requests that use the keycloak token will attempt to refresh it if it is about to be expired. 

Issue Number:  #342

## What is the new behavior?

Configuration options have been added for two major pieces of functionality -- the updateMinValidity seconds and update token exclusion http headers. When set, if one of the tokenUpdateExcludedHeaders matches one of the headers on an http request, the call to update the token will be skipped, simply adding the Bearer header and sending the request to the next handler. 

In tandem these two configuration options allow applications that passively poll authenticated endpoints frequently to only request a new token when the user is actively interacting with the application, thus allowing the user to be logged off automatically after a period of inactivity as defined by Keycloak.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

brainstain commented:
@mauriciovigolo  can we get a review of this addition please?  Enjoying your library, but need this as a way to enable polling without continual token renewal.

mauriciovigolo commented:
Hey @jeddai, I will be able to review it today. 
Thank you for your time and contribution ☺️

jonkoops commented:
I am wondering if this is a functionality that we want to include in core. Personally I am quite a fan of keeping things small and extensible to reduce the API complexity and maintenance burden. These kind of configuration based APIs tend to grow in complexity when more and more people try to add their own use-cases.

For example, what if I want to not update the token if a header is absent instead of present? Or what if I want to partially match a header with a regular expression? All of these use-cases (although imaginary right now) would require us to further extend the core API and ship that extra code to all other users which might never use these functionalities. 

I think that we should provide users with an API that they can easily extend to fit their use-cases. For example we can add an abstract method to determine if the request should actually add a token which can be overwritten to fit the exact use-case of the user.

jeddai commented:
> I am wondering if this is a functionality that we want to include in core. Personally I am quite a fan of keeping things small and extensible to reduce the API complexity and maintenance burden. These kind of configuration based APIs tend to grow in complexity when more and more people try to add their own use-cases.
>
> For example, what if I want to not update the token if a header is absent instead of present? Or what if I want to partially match a header with a regular expression? All of these use-cases (although imaginary right now) would require us to further extend the core API and ship that extra code to all other users which might never use these functionalities.
>
> I think that we should provide users with an API that they can easily extend to fit their use-cases. For example we can add an abstract method to determine if the request should actually add a token which can be overwritten to fit the exact use-case of the user.

@jonkoops that's a valid point. I'm more than happy to take a stab at implementing that if an extensible API would be preferable to the changes made here. I will note that the changes in this PR are specifically to prevent the token from being updated, not adding it to the request. There could, however, be separate hooks for adding the token to the request and allowing the token to be updated as a part of the request.

jonkoops commented:
That sounds like a great idea I would love to have some easy to override methods such as `shouldAddToken` and `shouldUpdateToken` so that these kind of rules could be implemented by users.

jeddai commented:
> That sounds like a great idea I would love to have some easy to override methods such as `shouldAddToken` and `shouldRenewToken` so that these kind of rules could be implemented by users.

I know we talked about this almost two weeks ago but I'm getting started on it today -- hope to have changes pushed in the next couple days.

jeddai commented:
> I think that we should provide users with an API that they can easily extend to fit their use-cases. 

@jonkoops @mauriciovigolo in the interest of implementing this while not introducing a breaking change, would it make sense to deprecate the `excludedUrls` option and remove it in a future release? The functionality offered by the `excludedUrls` option can be recreated by whatever application is using this library by passing a value for the `shouldAddToken` method.

jeddai commented:
Updated!

kevinahuber commented:
@mauriciovigolo Can you give this a review? It looks great and I'd like to start using it!

akyrey commented:
I agree this looks great! Whenever you got time, it would be great to be merged!

mauriciovigolo commented:
Hey @akyrey and @jeddai. 

I will have a closer look in this PR this week. I will save time for it!

Thanks again! ☺️

Waterstraal commented:
@mauriciovigolo Any ETA on this? Is there a way I can help?

jeddai commented:
I'll need to fix the merge conflicts, but it should be good to be reviewed as soon as I do.

jonkoops commented:
Apologies for the lack of reaction to this feature, I have been knee deep in my activities at Keycloak to really spend time on this project lately.

Waterstraal commented:
> I'll need to fix the merge conflicts, but it should be good to be reviewed as soon as I do.

@jeddai sounds good!

jeddai commented:
Fixed the merge conflicts, I think I noticed a good number of formatting changes, but my lint check finds like 47 warnings. Would you prefer I leave the warnings or should I run a lint fix?

jamesrweb commented on README.md:
```diff
@@ -186,9 +186,33 @@ export class AuthGuard extends KeycloakAuthGuard {
+    realm: 'your-realm',
+    clientId: 'your-client-id',
+  },
+  shouldAddToken: (req: HttpRequest<any>): Boolean => {
```
Since we know what should be in the request here, we should have a custom type such as:

```typescript
type ShouldAddTokenRequestPayload = {
  // ⬇ I am not sure if all these are possible, please proof and narrow the type accordingly! It would also be nice if there was a type we could use here such as KeycloakRequest["method"] to share the type definitions if something like that exists instead of using a custom union here.
  method: "get" | "post" | "put" | "delete",
  url: URL["pathname"]
}
```

Accordingly we can change the implementation to something like this:

```typescript
  // ⬇ "Boolean" is a JS constructor but "boolean" is a TS type
  shouldAddToken: (req: HttpRequest<ShouldAddTokenRequestPayload>): boolean => {
    try {
      const { method, url: path } = req;
      // ⬇ will throw if the path is not properly constructible as a valid url with the current origin and thus the `try / catch` block.
      const url = new URL(path, window.location.origin); 
      const acceptablePaths = ['/assets', '/clients/public'];
      const isGetRequest = method === "get";
      const isAcceptablePathMatch = acceptablePaths.some(pattern => pattern.test(url.pathname));
    
      return !(isGetRequest && urlMatch);
    } catch {
      return false;
    } 
  }
```

Comments in the above example should ofcourse be removed if used as they are intended for context purposes only.

WDYT?

jeddai replied:
Sounds good to me!

jeddai replied:
Reading through this more I'm a little confused by the intent of this. I can understand wanting the user to be able to understand what the `HttpRequest` object will have at a glance, but that type definition already exists within Angular's documentation. The generic being used here only determines the type of the request body (and some methods that relate to it), so I don't think adding a `ShouldAddTokenRequest` type would be beneficial.

As for updating the variable names, I'll adjust those. I'm not the best at choosing variable names haha.

jeddai replied:
Considering other readme sections include the relevant imports, maybe it would make sense to include the import for `HttpRequest`?

jamesrweb replied:
> Considering other readme sections include the relevant imports, maybe it would make sense to include the import for `HttpRequest`?

Yes, I agree, that would be nice.

jamesrweb replied:
> Reading through this more I'm a little confused by the intent of this. I can understand wanting the user to be able to understand what the `HttpRequest` object will have at a glance, but that type definition already exists within Angular's documentation. The generic being used here only determines the type of the request body (and some methods that relate to it), so I don't think adding a `ShouldAddTokenRequest` type would be beneficial.
> 
> As for updating the variable names, I'll adjust those. I'm not the best at choosing variable names haha.

Ah, I see, if we know the shape of the body then we could create a custom type for that then but if not, `unknown` over `any` might be nice to force a type guard validation on the users part. I don't know, you can decide for that one.

The two hardest things in computer science: Cache invalidation, naming things and off by one errors 😉!

jeddai replied:
`unknown` is a great idea, I'll update that!

jamesrweb commented on README.md:
```diff
@@ -197,7 +221,10 @@ await keycloak.init({
   },
-  bearerExcludedUrls: ['/assets', '/clients/public']
+  bearerExcludedUrls: ['/assets', '/clients/public'],
+  shouldUpdateToken: (req: HttpRequest<any>): Boolean => {
+    return !req.headers.get('token-update') === 'false';
+  }
```
Here we can similarly add stronger typing to the incoming request:

```typescript
type ShouldUpdateTokenRequest = {
  headers: Headers
}
```

```suggestion
  shouldUpdateToken: (req: HttpRequest<ShouldUpdateTokenRequest>): Boolean => {
    return !req.headers.get('token-update') === 'false';
  }
```

jamesrweb commented on projects/keycloak-angular/src/lib/core/services/keycloak.service.ts:
```diff
@@ -68,6 +69,18 @@ export class KeycloakService {
+   * The amount of required time remaining before expiry of the token before the token will be refreshed.
+   */
+  private _updateMinValidity: number;
+  /**
+   * Returns true if the request should have the token added to the headers by the KeycloakBearerInterceptor.
+   */
+  shouldAddToken: (_: HttpRequest<any>) => Boolean;
+  /**
+   * Returns true if the request being made should potentially update the token.
+   */
+  shouldUpdateToken: (_: HttpRequest<any>) => Boolean;
```
We can also add much stricter types here for each of the functions by injecting the full definition so users get a better DX.

```suggestion
  /**
   * Returns true if the request should have the token added to the headers by the KeycloakBearerInterceptor.
   */
  shouldAddToken: (_: HttpRequest<any>) => Boolean;
  /**
   * Returns true if the request being made should potentially update the token.
   */
  shouldUpdateToken: (_: HttpRequest<any>) => Boolean;
```

jeddai replied:
Can you elaborate what you mean by that?

jamesrweb replied:
I mean that if we know the response body shape then we should create a custom type and set it instead of `any` so we get proper typing and IDE support, etc, does that make sense? If we cannot know the shape of the response body then `unknown` could be a route to be clear about that. I don't know. WDYT?

jamesrweb commented on projects/keycloak-angular/src/lib/core/services/keycloak.service.ts:
```diff
@@ -162,14 +175,20 @@ export class KeycloakService {
-    initOptions
+    initOptions,
+    updateMinValidity = 20,
+    shouldAddToken = (_: HttpRequest<any>) => true,
+    shouldUpdateToken = (_: HttpRequest<any>) => true
```
Same here.

jeddai commented:
I'll try to get those changes made tonight!

jeddai commented:
So I made some of the suggested changes to the docs, in particular I added an import statement for the `HttpRequest` class to note that it's something that is imported from Angular.

I also updated a couple uses of `Boolean` in `keycloak.service.ts` to be `boolean` instead.

And lastly I ran prettier to fix the linting issues, I did notice that I think there might be some CLRF differences in the remote vs my local, so I hope that's not an issue but please let me know if it is.

jamesrweb commented:
Really great work @jeddai!

mauriciovigolo commented:
Hi @jeddai, the changes look great! It is a very nice add-on and I appreciated your work and support. Thank you 😊
